vies_service.go

```diff
@@ -0,0 +1,127 @@
+package vat
+
+import (
+	"bytes"
+	"encoding/xml"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+)
+
+// LookupServiceInterface is an interface for the service that calls external services to validate VATs.
+type LookupServiceInterface interface {
+	Validate(vatNumber string) error
+}
+
+// viesService validates EU VAT numbers with the VIES service
+type viesService struct{}
+
+// Validate returns whether the given VAT number is valid or not
+func (s *viesService) Validate(vatNumber string) error {
+	if len(vatNumber) < 3 {
+		return ErrInvalidVATNumberFormat
+	}
+
+	res, err := s.lookup(s.getEnvelope(vatNumber))
+	if err != nil {
+		return ErrServiceUnavailable{Err: err}
+	}
+	defer func() {
+		_ = res.Body.Close()
+	}()
+
+	xmlRes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return ErrUnexpected{Err: err}
+	}
+
+	// check if response contains "INVALID_INPUT" string
+	if bytes.Contains(xmlRes, []byte("INVALID_INPUT")) {
+		return ErrInvalidVATNumberFormat
+	}
+
+	// check if response contains "MS_UNAVAILABLE" string
+	if bytes.Contains(xmlRes, []byte("MS_UNAVAILABLE")) ||
+		bytes.Contains(xmlRes, []byte("MS_MAX_CONCURRENT_REQ")) {
+		return ErrServiceUnavailable{Err: nil}
+	}
+
+	var rd struct {
+		XMLName xml.Name `xml:"Envelope"`
+		Soap    struct {
+			XMLName xml.Name `xml:"Body"`
+			Soap    struct {
+				XMLName     xml.Name `xml:"checkVatResponse"`
+				CountryCode string   `xml:"countryCode"`
+				VATNumber   string   `xml:"vatNumber"`
+				RequestDate string   `xml:"requestDate"` // 2015-03-06+01:00
+				Valid       bool     `xml:"valid"`
+				Name        string   `xml:"name"`
+				Address     string   `xml:"address"`
+			}
+		}
+	}
+	if err = xml.Unmarshal(xmlRes, &rd); err != nil {
+		return ErrUnexpected{Err: err}
```

Comment:
And here too - if the response isn't XML its likely some gateway timeout error or something similar. Either way, `ErrServiceUnavailable` feels like the better option here too 🤔

Reply:
Guess we can do that and if we run into problems with it we can modify it again later.